### PR TITLE
fix a bug of encodeKey of DecimalType

### PR DESCRIPTION
**Pull request by zhexuany**
### What problem does this PR solve? 
When we encodeKey of DecimalType, we did not consider type's precision and fraction.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/sql/expression/index/UnsignedTestSuite.scala [hunk 1]
```diff
@@ -19,7 +19,8 @@ import org.apache.spark.sql.BaseTiSparkTest
 
 class UnsignedTestSuite extends BaseTiSparkTest {
 
-  test("Unsigned Index Tests for TISPARK-28 and TISPARK-29") {
+  // after we push down decimal type, unsigned long push-down logic need revisit.
```
Comment (birdstorm) [inline core/src/test/scala/org/apache/spark/sql/expression/index/UnsignedTestSuite.scala]: add a TODO?

**Comment by marsishandsome** [comment]
need to cherry-pick to release branch?

**Comment by zhexuany** [comment]
@marsishandsome This PR is merged to another branch rather than master. We can cherry pick it another branch later.

**Pull request merged**